### DISCUSSION TITLE
Custom id generator

### DISCUSSION THREAD

**Discussion by fairking** [go-rel/rel#326] [Q&A]
Hi guys,

Thanks for the wonderful package.
I have one small question. Is there any way to populate id automatically (on insert) if I have my own custom function to generate id?
So far I have read [this](https://go-rel.github.io/basics/#primary-key) and [this](https://go-rel.github.io/mutations/).

I know I can create a new struct value and populate the id manually:
```go
type Book struct {
    Id string
    Title string
}

book := Book{ Id: helper.GenerateId() }

repo.Insert(ctx, &book)
```

But maybe there is a way to tell REL to use my generator to populate the id on insert? So it would look like this:
```go
book := Book{}

repo.Insert(ctx, &book) // id generated automatically
```

Also I guess I can do this:
```go
book := Book{}

repo.Insert(ctx, &book, rel.Set("id", helper.GenerateId()))
```
But I feel it is too noisy.

The only solution I can see is to provide an extension method:
```go
func Insert(repo rel.Repository, ctx context, model interface{}) {
    repo.Insert(ctx, &model, rel.Set("id", helper.GenerateId()))
}

Insert(repo, ctx, model)
```
However I am not sure how relation would work, when `Book` struct has an `Author` field which is also a new value.

**Comment by Fs02** [comment]
above workaround is the only way for now, there's no functionality to add default id generation at the moment
